caffe2/contrib/tensorrt/tensorrt_op_trt.cc

```diff
@@ -0,0 +1,224 @@
+  int idx = 0;
+  for (const auto& oname : output_size_names) {
+    const auto it = outputs.find(oname);
+    if (it != outputs.end()) {
+      std::vector<TIndex> dims;
+      for (; idx < output_size_hints_encoded.size() && output_size_hints_encoded[idx] > 0; ++idx) {
+        dims.push_back(output_size_hints_encoded[idx]);
+      }
+      output_size_hints_.emplace(it->second, std::move(dims));
+    }
+  }
+
+  // match and bind the input/output
+  int num_bindings = trt_engine_->getNbBindings();
+  for (int b = 0; b < num_bindings; ++b) {
+    const auto& name = trt_engine_->getBindingName(b);
+    nv_dims_.push_back(trt_engine_->getBindingDimensions(b));
+    if (trt_engine_->bindingIsInput(b)) {
+      const auto it = inputs.find(name);
+      CAFFE_ENFORCE(it != inputs.end(), MakeString("Cannot find trt input: ", name));
+      binding_hints_.emplace_back(it->second, true);
+    } else {
+      const auto it = outputs.find(name);
+      CAFFE_ENFORCE(it != outputs.end());
+      binding_hints_.emplace_back(it->second, false);
+    }
+  }
+
+  trt_executor_ = InferObject(trt_engine_->createExecutionContext());
+}
+
+void TensorRTOp::MaybeAdjustOutputShape(int output_idx, std::vector<TIndex>* dims) {
+  const auto it = output_size_hints_.find(output_idx);
+  if (it != output_size_hints_.end()) {
+    const auto& dims_hint = it->second;
+    auto total_trt = std::accumulate(dims->begin(), dims->end(), (TIndex)(1), std::multiplies<TIndex>());
+    auto total_c2 = std::accumulate(dims_hint.begin(), dims_hint.end(), (TIndex)(1), std::multiplies<TIndex>());
+    if (total_c2 != total_trt) {
+      LOG(WARNING) << "The total size of TensorRT op output and hint don't match: " << total_trt << " vs " << total_c2;
+      return;
+    }
+
+    bool identical_shape = true;
+    if (dims->size() != dims_hint.size()) {
+      identical_shape = false;
+    } else {
+      for (int i = 0; i < dims->size(); ++i) {
+        if((*dims)[i] != dims_hint[i]) {
+          identical_shape = false;
+          break;
+        }
+      }
+    }
+
+    // We conform to the output shape hints. NB: We might need an explicit reshape op for this
+    if (!identical_shape) {
+      *dims = dims_hint;
+    }
+  }
+}
+
+bool TensorRTOp::RunOnDevice() {
+  CAFFE_ENFORCE(trt_executor_);
+  // Decide input batch size
+  size_t N = 0;
+  bool first = true;
+  for (int i = 0; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    if (first) {
+      N = tensor_dims.front();
+      first = false;
+    } else {
+      CAFFE_ENFORCE_EQ(
+          N, tensor_dims.front(), "Mismatched batch size in input tensors");
+    }
+  }
+
+  // We need to do the binding at RunOnDevice time because we only know the
+  // exact shapes of the tensors now. In addtion, since TensorRT engine has
+  // max_batch_size, we need to call that multiple times if input batch size
+  // exceeeds this limit.
```

Comment:
Could we avoid re-binding if shapes didn't change? A separate PR is fine

Reply:
The tricky part is that we cannot tell if we need rebinding just by comparing the shapes. If the input batch size is larger than `max_batch_size` we need to do multiple bindings in one device run. So the bindings changes between op calls even if the shapes are the same.

Reply:
I think binding is not costly. But if batch size exceeds `max_batch_size_` if makes sense to error out because calling the trt engine multiple times is not good.